parlai/agents/bert_ranker/biencoder_ranker.py

```diff
@@ -0,0 +1,194 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2017-present, Facebook, Inc.
+# All rights reserved.
+# This source code is licensed under the BSD-style license found in the
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+from parlai.core.torch_ranker_agent import TorchRankerAgent
+from .bert_dictionary import BertDictionaryAgent
+from .helpers import get_bert_optimizer, BertWrapper, BertModel, \
+    add_common_args, surround
+from parlai.core.utils import padded_3d
+import torch
+import json
+import tqdm
+
+
+class BiEncoderRankerAgent(TorchRankerAgent):
+    """ TorchRankerAgent implementation of the biencoder.
+        It is a standalone Agent. It might be called by the Both Encoder.
+    """
+
+    @staticmethod
+    def add_cmdline_args(parser):
+        add_common_args(parser)
+
+    def __init__(self, opt, shared=None):
+        opt['rank_candidates'] = True
+        opt["bert_id"] = 'bert-base-uncased'
+        opt['candidates'] = "batch"
+        if opt.get('eval_candidates', None) is None:
+            opt['eval_candidates'] = "inline"
+        self.clip = -1
+        super().__init__(opt, shared)
+        # NOTE: This is done AFTER init so that it's after load on purpose.
+        # the state dict of MyModule and DataParallel(MyModule) is not the same
+        if self.opt["multigpu"]:
+            self.model = torch.nn.DataParallel(self.model)
+        self.NULL_IDX = self.dict.pad_idx
+        self.START_IDX = self.dict.start_idx
+        self.END_IDX = self.dict.end_idx
+        # default one does not average
+        self.rank_loss = torch.nn.CrossEntropyLoss(reduce=True, size_average=True)
+
+    def build_model(self):
+        self.model = BiEncoderModule(self.opt)
+
+    @staticmethod
+    def dictionary_class():
+        return BertDictionaryAgent
+
+    def init_optim(self, params, optim_states=None, saved_optim_type=None):
+        if all(f in self.opt for f in ["num_samples", "num_epochs", "batchsize"]):
+            total_iterations = self.opt["num_samples"] * self.opt["num_epochs"] \
+                / self.opt["batchsize"]
+            self.optimizer = get_bert_optimizer([self.model],
+                                                self.opt["type_optimization"],
+                                                total_iterations,
+                                                0.05,  # 5% scheduled warmup.
+                                                self.opt["learningrate"])
+
+    def receive_metrics(self, metrics_dict):
+        """ Inibiting the scheduler.
+        """
+        pass
```

Comment:
:( I'm still not sure what's the right way for this, but I need to do something better in TA/TRA.

Reply:
Having an init_scheduler method that one can override to do nothing?

Reply:
the scheduler params in TA should be made into command line args. that way you can just override the factor to "1" so that the learning rate won't be reduced